modules/jdbc/src/main/java/org/apache/ignite/internal/jdbc/JdbcStatement.java

```diff
@@ -472,7 +462,7 @@ public boolean getMoreResults(int curr) throws SQLException {
             switch (curr) {
                 case CLOSE_CURRENT_RESULT:
                     if (curRes > 0) {
-                        resSets.get(curRes - 1).close0();
+                        resSets.get(curRes - 1).close0(false);
```

Comment:
Why can't we call `close0(true)` here?

The following example demonstrate resource leakage (or may be I missed something):
```
    // test can be added to ItJdbcMultiStatementSelfTest
    @Test
    public void testX() throws Exception {
        long initial = openCursorsRegistered();
        
        stmt.executeUpdate("CREATE TABLE TESTX(ID INT PRIMARY KEY)");
        stmt.getMoreResults();
        stmt.close();

        assertEquals(0L, initial - openCursorsRegistered());
    }
```

As I see, the current combination in `close0` of `isQuery`, `finished`, etc. flags were being used because we were not registering cursors for DML, DDL (and possibly other) queries.

Since we are now registering cursors for this types of queries, I think it is worth reviewing the current conditions in `close0`, because now they look too complex and leading to errors.

Reply:
this is connection context , append below to test and it will pass:
```
        conn.close();
        assertTrue(waitForCondition(() -> 0L == openCursorsRegistered(), 5000));
```

Reply:
Yes you are right. I missed this.
But as I see, now - getNextResultSet() already closes resultset with enforce=true, why we need this second call here?

Reply:
done